feat: add linter checks for naming conventions

deusebio opened the pull request:
This PR provides extra checks for compliance of configuration options, action names and action parameters with respect to use hyphens instead of snake case when composite/compound keys are used. 

The checks follows [this guideline](https://juju.is/docs/sdk/styleguide#heading--naming).

In general, we raise warning, although I believe that charms should provide strong consistency on the naming scheme adopted within the charm. So if one uses mixed contentions (some keys have hyphens and others have snake case), I would believe it is fair to error out. This could be by passed by using `--force`, but that would produce a strong nudge on charmers to NOT mix the two conventions.

lengau commented on charmcraft/linters.py:
```diff
@@ -325,10 +347,42 @@ def run(self, basedir: pathlib.Path) -> str:
 
+        actions_names = list(content.keys())
+
+        result_actions, text_actions = check_naming_convention(actions_names, "actions")
```
I think I'd rather have a separate naming conventions linter here that covers both the actions and the config items. Then you can use the `url` field for the link to the style guide, and it can be used to cover other things (e.g. charm names).

lengau reviewed (changes requested):
Thanks for the PR! I like the principle here (and would like to expand it).

Please redo this as a single (new) linter though.

deusebio commented on charmcraft/linters.py:
```diff
@@ -325,10 +347,42 @@ def run(self, basedir: pathlib.Path) -> str:
+
+        result_params, text_params = check_naming_convention(actions_params, "action params")
+
+        if result_actions == self.Result.ERROR or result_params == self.Result.ERROR:
```
All right! That's fine. I personally feel that people should really not mix naming conventions, but - really - this is your call. Happy to revert to WARNING for everything then.

deusebio replied:
BTW, if all the linting checks should all go into WARNING, we can probably simplify the interface of `check_naming_convention` a bit, with the function to just return True/False. If False we raise warning, otherwise we provide `Result.SUCCESS`

What do you think?

lengau replied:
From a linter perspective, I agree with you and do consider it an error. Unfortunately, with the current way linters in Charmcraft are set up, an error from a linter prevents completion of the charm build. So `WARNING` is more like a linter error in practice.

What about returning an `Optional[str]` from `check_naming_convention`? Then it can return `None` for success or a warning message for failure.

deusebio commented on charmcraft/linters.py:
```diff
@@ -325,10 +347,42 @@ def run(self, basedir: pathlib.Path) -> str:
         try:
             with filepath.open("rt", encoding="utf8") as fh:
-                yaml.safe_load(fh)
+                content = dict(yaml.safe_load(fh))
```
Oh I see. Sure! Thanks for flagging this. It was indeed subtle

lengau replied:
🦆 Duck typing! 🦆 

Fortunately if we move these changes into their own linter there's an easier way to handle it.

cmatsuoka commented on charmcraft/linters.py:
```diff
@@ -67,6 +68,27 @@ def check_dispatch_with_python_entrypoint(basedir: pathlib.Path) -> pathlib.Path
+def check_naming_convention(names: typing.Iterable[str], scope: str) -> tuple[str, str]:
+    snake_keys = [key for key in names if "_" in key]
+
+    if snake_keys:
+        hyphen_keys = [key for key in names if "-" in key]
```
@lengau is there a clear preference for one or the other? If we're warning against mixed use, should we also steer towards the preferred one?

deusebio replied:
@cmatsuoka The preferred way is `-` following the [styleguide](https://juju.is/docs/sdk/styleguide#heading--naming)

deusebio commented on tests/test_linters.py:
```diff
@@ -898,6 +900,63 @@ def test_jujuactions_file_corrupted(tmp_path):
     assert result == JujuActions.Result.ERROR
 
 
+def test_jujuactions_snake_case_action_params(tmp_path):
```
sure thing! I did not see other tests with this in the file so I did not want to introduce it myself :)

deusebio commented:
@lengau I refactored the code a bit hopefully in the direction of what you meant. I created one new Linter class where I centralized the logic for BOTH actions and config. Let me know if this is what you meant and I'll work out to fix other things (tests fails locally because of `Package backend 'apt' is not supported on this environment.`, not sure exactly if that's a regression I introduced as it seems unrelated)

lengau commented on charmcraft/linters.py:
```diff
@@ -376,6 +377,106 @@ def run(self, basedir: pathlib.Path) -> str:
+            return warnings
+
+        with config_file.open("rt", encoding="utf8") as fh:
+            options = yaml.safe_load(fh).get("options")
```
```suggestion
            options = yaml.safe_load(fh).get("options", {})
```

The user will get an error from the ConfigLinter before this runs, but this will still run so we should ensure it can handle invalid files.

Likewise both here and in `_actions_check`, please handle the cases where the file is empty (`safe_load` returns `None`) and where it's not valid YAML. We don't need to produce warnings again, we can just silently return. (And please test for these cases)

deusebio replied:
I believe in the actions this is already checked. Anyhow, I have added a test case to make sure that this was indeed true.

lengau reviewed (changes requested):
Thanks, this is great! With the minor changes I've requested and fixing the linting issues, I'll be happy to merge this.

deusebio commented:
strange that the "spread" tests did not pass as it used to :shrug:  Is it flaky on CI? @lengau: Could you just try to retrigger it?

lengau reviewed (approved):
Thanks!

Yes, the spread tests are unfortunately a bit flaky. We've got some open issues for solving them
